### PR TITLE
Physics ur

### PR DESCRIPTION
**Pull request by yuokamoto**
- Implement RRPhysicsJointComponent and Physics_UR10
  - Create PhysicsConstraint in constructor and set parameter in BeginPlay
  - Set pose and velocity target to PhysicsConstraints
  - In TickComponent, check reach goal or not
  - Add ability to manually set constraint directly in case overwrite by BP
  - if bSmooth is set, use two point interpolation to calc Pose and velocity
- Separate SetVelocityTarget and SetVelocity since physics can't set velocity directly
- Add joint valication
- Separate position and orientation tolerance

### CODE REVIEW

File: Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h [hunk 1]
```diff
@@ -166,15 +200,21 @@ class RAPYUTASIMULATIONPLUGINS_API URRJointComponent : public USceneComponent
...
-    //! Angular Velocity Limitations[deg/s]
     UPROPERTY(EditAnywhere, BlueprintReadWrite)
-    FVector AngularVelMin = FVector(-180, -180, -180);
+    bool IsLimitRoll = true;
+
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    bool IsLimitPitch = true;
+
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    bool IsLimitYaw = true;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h]:
```suggestion
    bool bLimitRoll = true;

    UPROPERTY(EditAnywhere, BlueprintReadWrite)
    bool bLimitPitch = true;

    UPROPERTY(EditAnywhere, BlueprintReadWrite)
    bool bLimitYaw = true;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h [hunk 2]
```diff
@@ -166,15 +200,21 @@ class RAPYUTASIMULATIONPLUGINS_API URRJointComponent : public USceneComponent
...
+
+protected:
+
+    FTransform JointToChildLink;
+    FTransform ParentLinkToJoint;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h]:
```suggestion
    UPROPERTY()
    FTransform JointToChildLink = FTransform::Identity;
    UPROPERTY()
    FTransform ParentLinkToJoint = FTransform::Identity;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 3]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! Smoothing TargetPose to #Constraint.
+    //! If this is false, step pose target are used by #SetPoseTarget
+    //! If this is true, pose target changes linearly with max vel in #TickComponent
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    bool bSmoothing = false;
+
+    //! Acceleration[cm/ss] used by velocity smoothing if #bVelocitySmoothing = true.
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float LinearVelocitySmoothingAcc = 10;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]:
```suggestion
    float LinearVelocitySmoothingAcc = 10.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 4]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! Acceleration[deg/ss] used by velocity smoothing if #bVelocitySmoothing = true.
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float AngularVelocitySmoothingAcc = 10;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]:
```suggestion
    float AngularVelocitySmoothingAcc = 10.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 5]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [cm] tolerance for control
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float PositionTolerance = 1;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]:
```suggestion
    float PositionTolerance = 1.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 6]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [degree] tolerance for control
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float OrientationTolerance = 1;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]:
```suggestion
    float OrientationTolerance = 1.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 7]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [cm/s] tolerance for control
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float LinearVelocityTolerance = 10;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]:
```suggestion
    float LinearVelocityTolerance = 10.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 8]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+    
+    //! [degree/s] tolerance for control
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float AngularVelocityTolerance = 10;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]:
```suggestion
    float AngularVelocityTolerance = 10.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 9]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [kg/ss]
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float LinearSpring = 1000;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h:129]:
```suggestion
    float LinearSpring = 1000.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 10]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [kg/s]
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float LinearBamper = 10;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h:133]:
```suggestion
    float LinearBamper = 10.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 11]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [kg.cm/ss]
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float LinearForceLimit = UE_BIG_NUMBER;
+
+    //! [kg.cm/ss]
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float AngularSpring = 100000;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h:141]:
```suggestion
    float AngularSpring = 100000.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 12]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+
+    //! [kg.cm/s]
+    UPROPERTY(EditAnywhere, BlueprintReadWrite)
+    float AngularBamper = 10000;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h:145]:
```suggestion
    float AngularBamper = 10000.f;
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h [hunk 13]
```diff
@@ -144,15 +187,6 @@ class RAPYUTASIMULATIONPLUGINS_API URRJointComponent : public USceneComponent
     UPROPERTY(EditAnywhere, BlueprintReadWrite)
     FVector PositionMin = FVector(-1000, -1000, -1000);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h:188]:
```suggestion
    FVector PositionMin = FVector(-1000.f);
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h [hunk 14]
```diff
@@ -166,15 +200,21 @@ class RAPYUTASIMULATIONPLUGINS_API URRJointComponent : public USceneComponent
     UPROPERTY(EditAnywhere, BlueprintReadWrite)
     FVector LinearVelMax = FVector(1000, 1000, 1000);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h:201]:
```suggestion
    FVector LinearVelMax = FVector(1000.f);
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h [hunk 15]
```diff
@@ -144,15 +187,6 @@ class RAPYUTASIMULATIONPLUGINS_API URRJointComponent : public USceneComponent
...
-
-    UPROPERTY(EditAnywhere, BlueprintReadWrite)
-    bool IsLimitPitch = true;
-
-    UPROPERTY(EditAnywhere, BlueprintReadWrite)
-    bool IsLimitYaw = true;
-
     //! Orientation Limitations[deg]
     UPROPERTY(EditAnywhere, BlueprintReadWrite)
     FRotator OrientationMax = FRotator(180, 180, 180);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h]:
```suggestion
    FRotator OrientationMax = FRotator(180.f);
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h [hunk 16]
```diff
@@ -166,15 +200,21 @@ class RAPYUTASIMULATIONPLUGINS_API URRJointComponent : public USceneComponent
     UPROPERTY(EditAnywhere, BlueprintReadWrite)
     FVector LinearVelMax = FVector(1000, 1000, 1000);
 
-    //! Linear Velocity Limitations[cm/s]
-    UPROPERTY(EditAnywhere, BlueprintReadWrite)
-    FVector LinearVelMin = FVector(-1000, -1000, -1000);
-
     //! Angular Velocity Limitations[deg/s]
     UPROPERTY(EditAnywhere, BlueprintReadWrite)
     FVector AngularVelMax = FVector(180, 180, 180);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRJointComponent.h]:
```suggestion
    FVector AngularVelMax = FVector(180.f);
```

File: Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h [hunk 17]
```diff
@@ -51,6 +90,87 @@ class RAPYUTASIMULATIONPLUGINS_API URRPhysicsJointComponent : public URRJointCom
...
+    FVector MidLinearVelocityTarget = FVector::ZeroVector;
+
+    //! AngularVelocityTarget for smoothing
+    FVector MidAngularVelocityTarget = FVector::ZeroVector;
+
+    //! PositionTarget for smoothing
+    FVector MidPositionTarget = FVector::ZeroVector;
+
+    //! OrientationTarget for smoothing
+    FRotator MidOrientationTarget = FRotator::ZeroRotator;
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Public/Drives/RRPhysicsJointComponent.h]: I used to experience non-reset member values between PIE->Stop->PIE->.. if without `UPROPERTY()` long ago in UE4, but might have been fixed by UE5 now

File: Source/RapyutaSimulationPlugins/RapyutaSimulationPlugins.Build.cs [hunk 18]
```diff
@@ -113,5 +113,10 @@ public RapyutaSimulationPlugins(ReadOnlyTargetRules Target) : base(Target)
         {
             AddLib(Target, SdformatPath, "libsdformat12.so", "libsdformat12.so.12");
         }
+
+        // two_point_interpolation
+        string TwoPointsInterpolationPath = GetLibPath("two_points_interpolation_cpp");
+        PublicIncludePaths.Add(TwoPointsInterpolationPath);
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/RapyutaSimulationPlugins.Build.cs:119]: If the method is not so big a lib, could we just have it added directly into RRMathUtils?

File: Source/RapyutaSimulationPlugins/Private/Drives/RRPhysicsJointComponent.cpp [hunk 19]
```diff
@@ -5,36 +5,381 @@
...
+                ChildLink->GetComponentTransform());
+        ParentLinkToJoint = URRGeneralUtils::GetRelativeTransform(
+                ParentLink->GetComponentTransform(),
+                Constraint->GetComponentTransform());
+        
+        for (uint8 i = 0; i < 3; i++)
+        {
+            PositionTPI[i] = TwoPointInterpolation(true);
+            OrientationTPI[i] = TwoAngleInterpolation(true);
+        }
```
Comment (Tadinu) [inline Source/RapyutaSimulationPlugins/Private/Drives/RRPhysicsJointComponent.cpp]:
```suggestion
        PositionTPI = decltype(PositionTPI)(InPlace, TwoPointInterpolation(true));
        OrientationTPI = decltype(OrientationTPI)(InPlace, TwoAngleInterpolation(true));
```

**Review by Tadinu — APPROVED** [review]
`float` const requires `.f` otherwise it will be stored as `double`

File: Source/RapyutaSimulationPlugins/RapyutaSimulationPlugins.Build.cs (hunk 18 above)
Comment (yuokamoto) [reply Source/RapyutaSimulationPlugins/RapyutaSimulationPlugins.Build.cs:119]: it is just one file header lib but want to keep in thirdparty but I can avoid submodule and directly add file.

Comment (Tadinu) [reply Source/RapyutaSimulationPlugins/RapyutaSimulationPlugins.Build.cs:119]: Yep, it'd depend on how you might expect the lib to grow, so up to you then. Inline adding the src to our UE plugin would have to come with a catch of possible future certain dependencies & coding convention, which are not always easily assured.

Comment (Tadinu) [reply Source/RapyutaSimulationPlugins/RapyutaSimulationPlugins.Build.cs:119]: Yep, it'd depend on how you might expect the lib to grow, so up to you then. Inline adding the src to our UE plugin could come with a catch of possible future certain dependencies & coding convention, which are not always easily assured.

**Comment by Tadinu** [comment]
I forgot to check, this seems to have missed .gitmodules file like this
https://github.com/rapyuta-robotics/RapyutaSimulationInternal/blob/devel/.gitmodules

**Pull request merged**